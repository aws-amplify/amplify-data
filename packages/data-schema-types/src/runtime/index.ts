// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export {
  generateCustomQueriesProperty,
  generateCustomMutationsProperty,
  generateCustomSubscriptionsProperty,
  generateEnumsProperty,
  generateModelsProperty,
} from './internals';
export * from './types';
