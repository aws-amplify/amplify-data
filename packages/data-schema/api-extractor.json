{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/src/index.d.ts"
}
