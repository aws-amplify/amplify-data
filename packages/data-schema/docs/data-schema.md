<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
