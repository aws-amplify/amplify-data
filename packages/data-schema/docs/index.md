<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md)

## API Reference

## Packages

|  Package | Description |
|  --- | --- |
|  [@aws-amplify/data-schema](./data-schema.md) |  |

