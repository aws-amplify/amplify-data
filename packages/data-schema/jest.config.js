/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
  testPathIgnorePatterns: [
    ".test-d.ts", // don't directly execute these files. They'll be tested via their `test.ts` counterparts
    "/node_modules/",
    "utils.ts",
    "/fixtures/",
    "/MappedTypes/",
  ],
  coverageThreshold: {
		global: {
			branches: 72,
			functions: 86,
			lines: 85,
			statements: 85,
		},
	},
};
