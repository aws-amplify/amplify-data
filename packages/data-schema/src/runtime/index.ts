// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { isApiGraphQLConfig } from './internals/utils/runtimeTypeGuards/isApiGraphQLProviderConfig';
import {
  ClientExtensions,
  ClientExtensionsSSRCookies,
  ClientExtensionsSSRRequest,
} from './client';
import {
  generateCustomQueriesProperty,
  generateCustomMutationsProperty,
  generateCustomSubscriptionsProperty,
  generateEnumsProperty,
  generateModelsProperty,
} from './internals';
import { generateModelsProperty as generateModelsPropertyServer } from './internals/server';
import {
  BaseClient,
  ClientInternalsGetter,
  GraphQLProviderConfig,
} from './bridge-types';
export {
  generateCustomQueriesProperty,
  generateCustomMutationsProperty,
  generateCustomSubscriptionsProperty,
  generateEnumsProperty,
  generateModelsProperty,
  isApiGraphQLConfig,
  isGraphQLResponseWithErrors,
  isConfigureEventWithResourceConfig,
} from './internals';

// TODO: separate import path
export function addSchemaToClientWithInstance<T extends Record<any, any>>(
  client: BaseClient,
  params: any,
  // params: ServerClientGenerationParams & CommonPublicClientOptions,
  getInternals: ClientInternalsGetter,
): BaseClient &
  (ClientExtensionsSSRCookies<T> | ClientExtensionsSSRRequest<T>) {
  const apiGraphqlConfig = params.config?.API?.GraphQL;

  if (isApiGraphQLConfig(apiGraphqlConfig)) {
    (client as any).models = generateModelsPropertyServer<T>(
      client as any,
      params,
      getInternals,
    );
    (client as any).enums = generateEnumsProperty<T>(apiGraphqlConfig);
    (client as any).queries = generateCustomQueriesProperty<T>(
      client as any,
      apiGraphqlConfig,
      getInternals,
    );
    (client as any).mutations = generateCustomMutationsProperty<T>(
      client as any,
      apiGraphqlConfig,
      getInternals,
    );
  }

  return client as any;
}

export function addSchemaToClient<T extends Record<any, any> = never>(
  client: BaseClient,
  apiGraphqlConfig: GraphQLProviderConfig['GraphQL'],
  getInternals: ClientInternalsGetter,
): BaseClient & ClientExtensions<T> {
  (client as any).models = generateModelsProperty<T>(
    client as any,
    apiGraphqlConfig,
    getInternals,
  );
  (client as any).enums = generateEnumsProperty<T>(apiGraphqlConfig);
  (client as any).queries = generateCustomQueriesProperty<T>(
    client as any,
    apiGraphqlConfig,
    getInternals,
  );
  (client as any).mutations = generateCustomMutationsProperty<T>(
    client as any,
    apiGraphqlConfig,
    getInternals,
  );
  (client as any).subscriptions = generateCustomSubscriptionsProperty(
    client as any,
    apiGraphqlConfig,
    getInternals,
  );
  return client as any;
}
