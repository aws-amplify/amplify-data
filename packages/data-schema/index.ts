import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
