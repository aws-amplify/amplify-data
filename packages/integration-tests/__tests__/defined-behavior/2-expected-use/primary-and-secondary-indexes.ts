import { a, ClientSchema } from '@aws-amplify/data-schema';
import type {
  Expect,
  Equal,
  ExtractModelMeta,
  Prettify,
} from '@aws-amplify/data-schema-types';
import { Amplify } from 'aws-amplify';
import type {
  ListPkOptions,
  ModelPrimaryCompositeKeyInput,
  StringFilter,
  IndexQueryInput,
} from '../../../../data-schema/src/runtime/client/index';
import {
  buildAmplifyConfig,
  mockedGenerateClient,
  optionsAndHeaders,
  expectSchemaModelExcludes,
} from '../../utils';
import { ModelSortDirection } from '@aws-amplify/data-schema/runtime';

/**
 * Custom Identifier and Secondary Index permutations
 *
 * Implicit/default identifier ('id') is already covered in ./implicit-system-fields.ts
 *
 * TK
 *
 */

// TODO: migrate exhaustive PK type tests from ./integration-tests/__tests__/operation-params.test-d.ts

describe('Primary Indexes', () => {
  describe('Custom identifier single field SK', () => {
    const schema = a
      .schema({
        Model: a
          .model({
            pk: a.string().required(),
            sk1: a.string().required(),
          })
          .identifier(['pk', 'sk1']),
      })
      .authorization((allow) => allow.publicApiKey());

    type Schema = ClientSchema<typeof schema>;

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('the client schema type has the expected identifier IR shape', () => {
      type ResolvedIdentifierMeta = Prettify<
        ExtractModelMeta<Schema>['Model']['identifier']
      >;
      type ExpectedIdentifier = {
        pk: {
          pk: string;
        };
        sk: {
          sk1: string;
        };
        compositeSk: never;
      };

      type _AssertIdentifierType = Expect<
        Equal<ResolvedIdentifierMeta, ExpectedIdentifier>
      >;
    });

    test('the client schema type has the expected list PK params', () => {
      type ModelMeta = ExtractModelMeta<Schema>['Model'];

      type ListPk = ListPkOptions<ModelMeta, Record<string, string>>;

      type ExpectedPkParams = {
        pk?: string;
        sk1?: StringFilter;
        sortDirection?: ModelSortDirection | undefined;
      };

      type _AssertIdentifierType = Expect<Equal<ListPk, ExpectedPkParams>>;
    });

    test('the generated graphql excludes `id: ID!` PK', async () => {
      expectSchemaModelExcludes({
        schema: schema.transform().schema,
        model: 'Model',
        field: 'id',
      });
    });

    test('the generated modelIntrospection schema contains the expected PK and SK fields and metadata', async () => {
      expect.assertions(3);

      const { modelIntrospection } = await buildAmplifyConfig(schema);
      expect(modelIntrospection.models['Model'].primaryKeyInfo).toEqual(
        expect.objectContaining({
          isCustomPrimaryKey: true,
          primaryKeyFieldName: 'pk',
          sortKeyFieldNames: ['sk1'],
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['pk']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'pk',
          type: 'String',
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['sk1']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'sk1',
          type: 'String',
        }),
      );
    });

    test('the runtime graphql document has expected types for PK and SK fields', async () => {
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const { spy, generateClient } = mockedGenerateClient([
        { data: { listModels: { items: [] } } },
      ]);
      const client = generateClient<Schema>();
      await client.models.Model.list({
        pk: 'abc',
        sk1: {
          gt: 'a',
        },
      });

      const [[{ query, variables }]] = optionsAndHeaders(spy);

      expect(query).toEqual(
        expect.stringContaining(
          '$pk: String,$sortDirection: ModelSortDirection,$sk1: ModelStringKeyConditionInput',
        ),
      );

      expect(variables).toStrictEqual({
        pk: 'abc',
        sk1: {
          gt: 'a',
        },
      });
    });
  });

  describe('Custom identifier with composite SK', () => {
    const schema = a
      .schema({
        Model: a
          .model({
            pk: a.string().required(),
            sk1: a.string().required(),
            sk2: a.integer().required(),
          })
          .identifier(['pk', 'sk1', 'sk2']),
      })
      .authorization((allow) => allow.publicApiKey());

    type Schema = ClientSchema<typeof schema>;

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('the client schema type has the expected identifier IR shape', () => {
      type ResolvedIdentifierMeta = Prettify<
        ExtractModelMeta<Schema>['Model']['identifier']
      >;
      type ExpectedIdentifier = {
        pk: {
          pk: string;
        };
        sk: {
          sk1: string;
          sk2: number;
        };
        compositeSk: 'sk1Sk2';
      };

      type _AssertIdentifierType = Expect<
        Equal<ResolvedIdentifierMeta, ExpectedIdentifier>
      >;
    });

    test('the client schema type has the expected list PK params', () => {
      type ModelMeta = ExtractModelMeta<Schema>['Model'];

      type ListPk = ListPkOptions<ModelMeta, Record<string, string>>;

      type ExpectedPkParams = {
        pk?: string;
        sk1Sk2?: ModelPrimaryCompositeKeyInput<{ sk1: string; sk2: number }>;
        sortDirection?: ModelSortDirection | undefined;
      };

      type _AssertIdentifierType = Expect<Equal<ListPk, ExpectedPkParams>>;
    });

    test('the generated graphql excludes `id: ID!` PK', async () => {
      expectSchemaModelExcludes({
        schema: schema.transform().schema,
        model: 'Model',
        field: 'id',
      });
    });

    test('the generated modelIntrospection schema contains the expected PK and SK fields and metadata', async () => {
      expect.assertions(4);

      const { modelIntrospection } = await buildAmplifyConfig(schema);
      expect(modelIntrospection.models['Model'].primaryKeyInfo).toEqual(
        expect.objectContaining({
          isCustomPrimaryKey: true,
          primaryKeyFieldName: 'pk',
          sortKeyFieldNames: ['sk1', 'sk2'],
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['pk']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'pk',
          type: 'String',
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['sk1']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'sk1',
          type: 'String',
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['sk2']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'sk2',
          type: 'Int',
        }),
      );
    });

    test('the runtime graphql document has expected types for PK and SK fields', async () => {
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const { spy, generateClient } = mockedGenerateClient([
        { data: { listModels: { items: [] } } },
      ]);
      const client = generateClient<Schema>();
      await client.models.Model.list({
        pk: 'abc',
        sk1Sk2: {
          gt: {
            sk1: 'a',
            sk2: 1,
          },
        },
      });

      const [[{ query, variables }]] = optionsAndHeaders(spy);

      expect(query).toEqual(
        expect.stringContaining(
          '$pk: String,$sortDirection: ModelSortDirection,$sk1Sk2: ModelModelPrimaryCompositeKeyConditionInput',
        ),
      );

      expect(variables).toStrictEqual({
        pk: 'abc',
        sk1Sk2: {
          gt: {
            sk1: 'a',
            sk2: 1,
          },
        },
      });
    });
  });
});

describe('Secondary Indexes', () => {
  describe('Secondary index with single field SK', () => {
    const schema = a
      .schema({
        Model: a
          .model({
            gsiPk: a.string().required(),
            gsiSk: a.string().required(),
          })
          .secondaryIndexes((index) => [index('gsiPk').sortKeys(['gsiSk'])]),
      })
      .authorization((allow) => allow.publicApiKey());

    type Schema = ClientSchema<typeof schema>;

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('the client schema type has the expected secondary index IR shape', () => {
      type ResolvedGsiMeta = Prettify<
        ExtractModelMeta<Schema>['Model']['secondaryIndexes']
      >;
      type ExpectedGsiMeta = [
        {
          defaultQueryFieldSuffix: 'GsiPkAndGsiSk';
          queryField: never;
          pk: {
            gsiPk: string;
          };
          sk: {
            gsiSk: string;
          };
          compositeSk: never;
        },
      ];

      type _AssertIdentifierType = Expect<
        Equal<ResolvedGsiMeta, ExpectedGsiMeta>
      >;
    });

    test('the client schema type has the expected index query input params', () => {
      type ModelName = 'Model';
      type ModelMeta = ExtractModelMeta<Schema>[ModelName];

      type ResolvedIndexQueryInput = Prettify<
        IndexQueryInput<
          ModelMeta['secondaryIndexes'][0],
          Record<string, string>
        >
      >;

      type ExpectedIndexQueryInput = {
        gsiPk: string;
        gsiSk?: StringFilter;
      };

      type _AssertIdentifierType = Expect<
        Equal<ResolvedIndexQueryInput, ExpectedIndexQueryInput>
      >;
    });

    test('the generated modelIntrospection schema contains the expected index fields and key metadata', async () => {
      expect.assertions(3);

      const { modelIntrospection } = await buildAmplifyConfig(schema);

      const indexKeyAttribute = modelIntrospection.models[
        'Model'
      ].attributes.find(
        (attr: Record<string, any>) =>
          attr.type === 'key' &&
          attr.properties.queryField === 'listModelByGsiPkAndGsiSk',
      );

      expect(indexKeyAttribute).toEqual(
        expect.objectContaining({
          type: 'key',
          properties: {
            name: 'modelsByGsiPkAndGsiSk',
            queryField: 'listModelByGsiPkAndGsiSk',
            fields: ['gsiPk', 'gsiSk'],
          },
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['gsiPk']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'gsiPk',
          type: 'String',
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['gsiSk']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'gsiSk',
          type: 'String',
        }),
      );
    });

    test('the runtime graphql document has expected types for GSI PK and SK fields', async () => {
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const { spy, generateClient } = mockedGenerateClient([
        { data: { listModels: { items: [] } } },
      ]);
      const client = generateClient<Schema>();
      await client.models.Model.listModelByGsiPkAndGsiSk({
        gsiPk: 'abc',
        gsiSk: { beginsWith: '...' },
      });

      const [[{ query, variables }]] = optionsAndHeaders(spy);

      expect(query).toEqual(
        expect.stringContaining(
          '$gsiPk: String!,$gsiSk: ModelStringKeyConditionInput',
        ),
      );
      expect(query).toEqual(
        expect.stringContaining('$sortDirection: ModelSortDirection,'),
      );

      expect(variables).toStrictEqual({
        gsiPk: 'abc',
        gsiSk: { beginsWith: '...' },
      });
    });
  });
  describe('Secondary index with composite SK', () => {
    const schema = a
      .schema({
        Model: a
          .model({
            gsiPk: a.string().required(),
            gsiSk1: a.string().required(),
            gsiSk2: a.integer().required(),
          })
          .secondaryIndexes((index) => [
            index('gsiPk').sortKeys(['gsiSk1', 'gsiSk2']),
          ]),
      })
      .authorization((allow) => allow.publicApiKey());

    type Schema = ClientSchema<typeof schema>;

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('the client schema type has the expected secondary index IR shape', () => {
      type ResolvedGsiMeta = Prettify<
        ExtractModelMeta<Schema>['Model']['secondaryIndexes']
      >;
      type ExpectedGsiMeta = [
        {
          defaultQueryFieldSuffix: 'GsiPkAndGsiSk1AndGsiSk2';
          queryField: never;
          pk: {
            gsiPk: string;
          };
          sk: {
            gsiSk1: string;
            gsiSk2: number;
          };
          compositeSk: 'gsiSk1GsiSk2';
        },
      ];

      type _AssertIdentifierType = Expect<
        Equal<ResolvedGsiMeta, ExpectedGsiMeta>
      >;
    });

    test('the client schema type has the expected index query input params', () => {
      type ModelName = 'Model';
      type ModelMeta = ExtractModelMeta<Schema>[ModelName];

      type ResolvedIndexQueryInput = Prettify<
        IndexQueryInput<
          ModelMeta['secondaryIndexes'][0],
          Record<string, string>
        >
      >;

      type ExpectedIndexQueryInput = {
        gsiPk: string;
        gsiSk1GsiSk2?: ModelPrimaryCompositeKeyInput<{
          gsiSk1: string;
          gsiSk2: number;
        }>;
      };

      type _AssertIdentifierType = Expect<
        Equal<ResolvedIndexQueryInput, ExpectedIndexQueryInput>
      >;
    });

    test('the generated modelIntrospection schema contains the expected index fields and key metadata', async () => {
      expect.assertions(4);

      const { modelIntrospection } = await buildAmplifyConfig(schema);

      const indexKeyAttribute = modelIntrospection.models[
        'Model'
      ].attributes.find(
        (attr: Record<string, any>) =>
          attr.type === 'key' &&
          attr.properties.queryField === 'listModelByGsiPkAndGsiSk1AndGsiSk2',
      );

      expect(indexKeyAttribute).toEqual(
        expect.objectContaining({
          type: 'key',
          properties: {
            name: 'modelsByGsiPkAndGsiSk1AndGsiSk2',
            queryField: 'listModelByGsiPkAndGsiSk1AndGsiSk2',
            fields: ['gsiPk', 'gsiSk1', 'gsiSk2'],
          },
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['gsiPk']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'gsiPk',
          type: 'String',
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['gsiSk1']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'gsiSk1',
          type: 'String',
        }),
      );

      expect(modelIntrospection.models['Model']['fields']['gsiSk2']).toEqual(
        expect.objectContaining({
          isArray: false,
          isRequired: true,
          name: 'gsiSk2',
          type: 'Int',
        }),
      );
    });

    test('the runtime graphql document has expected types for GSI PK and SK fields', async () => {
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const { spy, generateClient } = mockedGenerateClient([
        { data: { listModels: { items: [] } } },
      ]);
      const client = generateClient<Schema>();
      await client.models.Model.listModelByGsiPkAndGsiSk1AndGsiSk2({
        gsiPk: 'abc',
        gsiSk1GsiSk2: { beginsWith: { gsiSk1: '...', gsiSk2: 0 } },
      });

      const [[{ query, variables }]] = optionsAndHeaders(spy);

      expect(query).toEqual(
        expect.stringContaining(
          '$gsiPk: String!,$gsiSk1GsiSk2: ModelModelModelsByGsiPkAndGsiSk1AndGsiSk2CompositeKeyConditionInput',
        ),
      );
      expect(query).toEqual(
        expect.stringContaining('$sortDirection: ModelSortDirection,'),
      );

      expect(variables).toStrictEqual({
        gsiPk: 'abc',
        gsiSk1GsiSk2: { beginsWith: { gsiSk1: '...', gsiSk2: 0 } },
      });
    });
  });
});
