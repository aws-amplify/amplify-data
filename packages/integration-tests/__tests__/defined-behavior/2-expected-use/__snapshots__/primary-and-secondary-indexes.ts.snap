// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primary Indexes Custom identifier single field SK defaulted/serial field can be used in identifier 1`] = `
"type Model @model @auth(rules: [{allow: public, provider: apiKey}])
{
  idSerial: Int! @primaryKey @default
  content: String
}"
`;
